roxie/udplib/udptrs.cpp

```diff
@@ -312,202 +405,62 @@ class CSendManager : implements ISendManager, public CInterface
         virtual int doRun() = 0;
     };
 
-
-    class send_send_flow : public StartedThread 
+    class send_resend_flow : public StartedThread
     {
-        /*
-        I don't like this code much at all
-        Looping round all every time status of any changes seems like a bad thing especially as scale goes up
-        Even though these look like a bitmap they are not used as such presently
-         - as a result, if data_added() is called while state is completed, we lose the request I think
-         - probably get away with it because of the dataqueued check
-
-        doRun() uses state bits without protection
-
-        A count of number pending for each might be better than just a flag
-
-        Circular buffers to give a list of which ones are in a given state would speed up the processing in the thread?
-         - normally MANY in pending (and order is interesting)
-         - normally few in any other state (only 1 if thread keeping up), order not really very interesting
-         - Want to keep expense on caller threads low (at the moment just set flag and sometimes signal)
-          - in particular don't lock while processing the chain
-         - Never need to be in >1 chain
-        msTick() probably better than time() for detecting timeouts
-
-        */
-        enum bits { new_request = 1, pending_request = 2, sending_data = 4, completed = 8, completed_more = 16 };
-
-        unsigned target_count;
-
-        char *state;
-        unsigned char *timeouts;   // Number of consecutive timeouts
-        unsigned *request_time;
-
-        CriticalSection cr;
-        Semaphore       sem;
+        // Check if any senders have timed out
         CSendManager &parent;
+        Semaphore terminated;
 
-        virtual int doRun() 
+        virtual int doRun() override
         {
-            // MORE - this is reading the state values unprotected
-            // Not sure that this represents any issue in practice...
             if (udpTraceLevel > 0)
-                DBGLOG("UdpSender: send_send_flow started - node=%u", parent.myNodeIndex);
-
-            while (running) 
+                DBGLOG("UdpSender: send_resend_flow started");
+            while (running)
             {
-                bool idle = false;
-                if (sem.wait(1000)) 
-                {
-                    if (udpTraceLevel > 4)
-                        DBGLOG("UdpSender: send_send_flow::doRun signal received");
-                }
-                else
-                    idle = true;
-                if (!running) return 0;
-
+                unsigned timeout = udpRequestToSendTimeout;
```

Comment:
This should be initialised before the loop, and then reset after the wait.  Otherwise the code for adjusting it on line 445 has no effect.

Reply:
Changed